### ISSUE TITLE
Add 'notrace' and 'static' attribute to the most of our internal functions

### ISSUE THREAD

**Issue by Adam-pi3** [lkrg-org/lkrg#21]

**Comment by solardiz** [comment]
Adam, please provide some rationale for this. Thanks!

**Comment by Adam-pi3** [comment]
Sure. When FTRACE is enabled GCC compiles every code with -pg option. It adds a stub at the beginning of each function (in fact long NOP) which can be dynamically "overwritten" and place 'jump' to the specified code to execute when needed. It also applies to LKRG's functions (including critical functions).
However, some of the functions in the kernel are very critical and we do not want to allow such functionality. Linux kernel to protect from that introduced 'notrace' attribute and such functions can't be hooked (e.g. FTRACE's internal functions to avoid deadlocks).
All available functions for hooking are listed in user-mode via FTRACE interface (e.g. /sys/kernel/debug/tracing/available_filter_functions). From FTRACE perspective they are called 'filters'. We can set-up custom code for 'filters' and set-up various actions like tracking their perf etc. However, dynamic ftrace (from kernel) can also modify their context (e.g. registers, including PC). In general it's a potential security issue for LKRG. We should avoid that feature.

**Comment by solardiz** [comment]
Adam, while at it please also be adding `static` to most functions. Thanks!

**Comment by solardiz** [comment]
Now we also have many functions in the embedded libhydrogen. It'd be nasty to make changes to those as that would make comparisons and sync'ing to latest upstream more difficult. I don't have a good suggestion on what to do about that. But until we get there, we still have plenty of LKRG-specific functions to add `notrace` to.

**Comment by solardiz** [comment]
It looks like more consistent use of `notrace` by us would break Singularity rootkit's current LKRG bypass, where it hooks our functions.

@kerneltoast suggests we could also try `ccflags-remove-y := $(CC_FLAGS_FTRACE)` instead of `notrace` in the source... but I guess it's better in addition, because some people integrate LKRG into the kernel perhaps without usage of our `Makefile`?

**Comment by kerneltoast** [comment]
> because some people integrate LKRG into the kernel perhaps without usage of our `Makefile`?

I think it counts as user error to integrate LKRG into the kernel without usage of `scripts/copy-builtin.sh`, which indeed generates a `Makefile` for LKRG.

**Comment by solardiz** [comment]
> I think it counts as user error to integrate LKRG into the kernel without usage of `scripts/copy-builtin.sh`

I agree. So we may want to add `ccflags-remove-y := $(CC_FLAGS_FTRACE)` to both our top-level `Makefile` and `scripts/copy-builtin.sh`.

@kerneltoast Do you want to give this a try and send us a PR, since this was your idea? Thank you!

**Comment by Adam-pi3** [comment]
@solardiz what about `hide` option, would it break the rootkit too? If yes maybe we should reconsider enabling it by default since we know it also improve other attacks on LKRG?

**Comment by solardiz** [comment]
@Adam-pi3 Yes, like I already said in chat with @kerneltoast "They hook LKRG functions by name. I guess setting our hide=1 sysctl will break that." But this is off-topic for this GitHub issue. We could want to open another.

**Comment by kerneltoast** [comment]
`ccflags-remove-y := $(CC_FLAGS_FTRACE)` indeed works; however, `ccflags-remove-y` was only introduced in 5.9. Also, `CC_FLAGS_FTRACE` was introduced in 4.0.

The only way to pull this off for all kernel versions supported by LKRG is to specify a `CFLAGS_REMOVE_<file>.o` line for every LKRG object. Thoughts @solardiz?

**Comment by solardiz** [comment]
@kerneltoast What happens on older kernels when we try to use `ccflags-remove-y`? Is it simply ignored? If so, partial extra protection for 5.9+ makes sense to have. "A `CFLAGS_REMOVE_<file>.o` line for every LKRG object" is also a fine thing to do, I think - would work for 4.0+, right?

**Comment by kerneltoast** [comment]
@solardiz `ccflags-remove-y` is a kbuild feature that was only added in 5.9; setting it on older kernels simply does nothing since the build system doesn't look for that variable at all. 

`CFLAGS_REMOVE_<file>.o` is supported since 2.6.27, so it can be used on all kernel versions supported by LKRG.

The `CC_FLAGS_FTRACE` variable only exists since 4.0, so to support older kernels than that, we'd simply have to manually specify the cflags to remove (e.g. `-pg`) instead of using the convenient `CC_FLAGS_FTRACE` variable.

**Comment by Adam-pi3** [comment]
> `CFLAGS_REMOVE_<file>.o` is supported since 2.6.27, so it can be used on all kernel versions supported by LKRG.

We should co with this path since we want to support 3.xx kernels
